### PR TITLE
qemu.tests: Add new case migration_multi_host_with_stress

### PR DESCRIPTION
**Pull request by spcui**
Signed-off-by: Shuping Cui <scui@redhat.com>

ID: 1256112

### CODE REVIEW

File: qemu/tests/cfg/multi_host.cfg [hunk 1]
```diff
@@ -108,6 +108,12 @@
                     #Count of migration during file transfer.
                     migrate_count = 3
                     regain_ip_cmd = ""
+                - with_stress:
+                    only Linux
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: Hi,
Where is the test for Windows?

File: qemu/tests/cfg/multi_host.cfg [hunk 2]
```diff
@@ -108,6 +108,12 @@
...
+                    only after_login_vm
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: need to update it, as after_login_vm will be moved to the same variant of other tests

File: qemu/tests/cfg/multi_host.cfg (hunk 1 above)
Comment (spcui) [inline qemu/tests/cfg/multi_host.cfg]: only support linux now

File: qemu/tests/cfg/multi_host.cfg (hunk 2 above)
Comment (spcui) [inline qemu/tests/cfg/multi_host.cfg]: yes, will send other PR to fix the config issue.

File: qemu/tests/migration_multi_host_with_stress.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        mig_type = utils_test.qemu.MultihostMigrationRdma
+
+    class TestMultihostMigration(mig_type):
+
+        def __init__(self, test, params, env):
+            super(TestMultihostMigration, self).__init__(test, params, env,
+                                                         preprocess_env)
+            self.srchost = self.params.get("hosts")[0]
+            self.dsthost = self.params.get("hosts")[1]
+            self.vm = params["vms"].split()[0]
```
Comment (suqinhuang) [inline qemu/tests/migration_multi_host_with_stress.py]: Hi scui,

Why just get first vm

File: qemu/tests/migration_multi_host_with_stress.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                vm = env.get_vm(params["main_vm"])
+                session = vm.wait_for_login(timeout=self.login_timeout)
+
+                error.context("Do stress test before migration.", logging.info)
+                if params.get("os_type") == "linux":
+                    bg = utils.InterruptedThread(self.do_stress, (session,))
+                else:
+                    bg = utils.InterruptedThread(utils_test.run_virt_sub_test,
+                                                 args=(test, params, env,),
+                                                 kwargs={"sub_type": "win_heavyload"})
```
Comment (suqinhuang) [inline qemu/tests/migration_multi_host_with_stress.py]: Could you set it in cfg file,
and some params is needed:
install_cmd, config_cmd, details pls check win_heavyload test case.

Comment (suqinhuang) [inline qemu/tests/migration_multi_host_with_stress.py]: How long will win_heavyload run?

File: qemu/tests/cfg/multi_host.cfg [hunk 5]
```diff
@@ -108,6 +108,11 @@
                     #Count of migration during file transfer.
                     migrate_count = 3
                     regain_ip_cmd = ""
+                - with_stress:
+                    only after_login_vm
+                    only mig_online
+                    type = migration_multi_host_with_stress
+                    stress_cmd = "stress --cpu 4 --io 4 --vm 2 --vm-bytes 128M --timeout 300s &"
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: will stress be installed ?

File: qemu/tests/migration_multi_host_with_stress.py (hunk 4 above)
Comment (spcui) [inline qemu/tests/migration_multi_host_with_stress.py]: timeout set in win_heavyload.cfg

File: qemu/tests/cfg/multi_host.cfg (hunk 5 above)
Comment (spcui) [inline qemu/tests/cfg/multi_host.cfg]: yes, already installed after installation linux guest.

File: qemu/tests/migration_multi_host_with_stress.py (hunk 3 above)
Comment (spcui) [inline qemu/tests/migration_multi_host_with_stress.py]: ok, will update it.

File: qemu/tests/cfg/multi_host.cfg [hunk 6]
```diff
@@ -108,6 +108,13 @@
...
+                    Windows:
+                        bg_stress_test = win_heavyload
                 - downtime:
                     only after_login_vm
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: Could you send patch to fix after_login_vm issue first?

File: qemu/tests/cfg/multi_host.cfg [hunk 7]
```diff
@@ -108,6 +108,13 @@
                     #Count of migration during file transfer.
                     migrate_count = 3
                     regain_ip_cmd = ""
+                - with_stress:
+                    only after_login_vm
+                    only mig_online
+                    type = migration_multi_host_with_stress
+                    stress_cmd = "stress --cpu 4 --io 4 --vm 2 --vm-bytes 128M --timeout 300s &"
+                    Windows:
+                        bg_stress_test = win_heavyload
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: no need params
install_path install_cmd config_cmd ?

Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: "how long will stress run in Windows ? " which I comment before

Comment (spcui) [inline qemu/tests/cfg/multi_host.cfg]: add related config

File: qemu/tests/migration_multi_host_with_stress.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            def worker(mig_data):
+                logging.info("Try to login guest before migration test.")
+                vm = env.get_vm(params["main_vm"])
+                session = vm.wait_for_login(timeout=self.login_timeout)
+
+                error.context("Do stress test before migration.", logging.info)
+                bg_stress_test = params.get("bg_stress_test")
+
+                if params.get("os_type") == "linux":
+                    bg = utils.InterruptedThread(self.do_stress, (session,))
```
Comment (suqinhuang) [inline qemu/tests/migration_multi_host_with_stress.py]: use autotest.stress

File: qemu/tests/migration_multi_host_with_stress.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+                else:
+                    bg = utils.InterruptedThread(utils_test.run_virt_sub_test,
+                                                 args=(test, params, env,),
+                                                 kwargs={"sub_type": bg_stress_test})
+
+                bg.start()
```
Comment (CongSmile) [inline qemu/tests/migration_multi_host_with_stress.py]: Need check the thread is finished or not.

File: qemu/tests/cfg/multi_host.cfg [hunk 10]
```diff
@@ -108,6 +108,20 @@
...
+                        autostress = yes
+                        install_cmd = "start /wait DRIVE:\HeavyLoadSetup.exe /verysilent"
+                        config_cmd = 'setx -m path "%PATH%;${install_path};"'
+                        i386,i686:
+                            install_path = "C:\Program Files\JAM Software\HeavyLoad"
+                        x86_64:
+                            install_path = "C:\Program Files (x86)\JAM Software\HeavyLoad"
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: pls move guest cfg to guest-os

**Comment by spcui** [comment]
deps on:
https://github.com/avocado-framework/avocado-vt/pull/295

File: qemu/tests/migration_multi_host_with_stress.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+                if not bg.is_alive():
+                    raise error.TestFail("Failed to start guest test!")
+                # waiting background test start.
+                time.sleep(wait_bg_timeout)
+
+            def check_worker(mig_data):
+                if not self.is_src and self.need_cleanup:
+                    vm = env.get_vm(params["main_vm"])
+                    clean_up(vm)
+                else:
```
Comment (xutian) [inline qemu/tests/migration_multi_host_with_stress.py]: useless else branch

File: qemu/tests/migration_multi_host_with_stress.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        4) login guest when migrate done
+
+    :param test: kvm test object.
+    :param params: Dictionary with test parameters.
+    :param env: Dictionary with the test environment.
+    """
+    preprocess_env = params.get("preprocess_env", "yes") == "yes"
+    mig_protocol = params.get("mig_protocol", "tcp")
+    mig_type = utils_test.qemu.MultihostMigration
+    if mig_protocol == "fd":
```
Comment (xutian) [inline qemu/tests/migration_multi_host_with_stress.py]: do we still need support offline migrate?

Comment (spcui) [inline qemu/tests/migration_multi_host_with_stress.py]: not support now.

File: qemu/tests/migration_multi_host_with_stress.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+            self.need_cleanup = params.get("need_cleanup", "yes") == "yes"
+
+        def migration_scenario(self):
+            def clean_up(vm):
+                kill_bg_stress_cmd = params.get("kill_bg_stress_cmd",
+                                                "killall -9 stress")
+
+                logging.info("Kill the background test in the guest.")
+                session = vm.wait_for_login(timeout=self.login_timeout)
+                session.sendline(kill_bg_stress_cmd)
```
Comment (xutian) [inline qemu/tests/migration_multi_host_with_stress.py]: sendline  will not check cmd status, so how to ensure stress really stopped?

File: qemu/tests/migration_multi_host_with_stress.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+                error.context("Do stress test before migration.", logging.info)
+                bg_stress_test = params.get("bg_stress_test")
+
+                bg = utils.InterruptedThread(utils_test.run_virt_sub_test,
+                                             args=(test, params, env,),
+                                             kwargs={"sub_type": bg_stress_test})
+
+                bg.start()
+                if not bg.is_alive():
+                    raise error.TestFail("Failed to start guest test!")
```
Comment (xutian) [inline qemu/tests/migration_multi_host_with_stress.py]: report which sub test start failed, looks more useful

File: qemu/tests/migration_multi_host_with_stress.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+                # waiting background test start.
+                time.sleep(wait_bg_timeout)
```
Comment (xutian) [inline qemu/tests/migration_multi_host_with_stress.py]: why not utils_misc.wait_for(bg.is_alive, timeout=wait_bg_timeout)?

File: qemu/tests/migration_multi_host_with_stress.py (hunk 12 above)
Comment (xutian) [inline qemu/tests/migration_multi_host_with_stress.py]: so remove useless code !!

File: qemu/tests/migration_multi_host_with_stress.py (hunk 15 above)
Comment (spcui) [inline qemu/tests/migration_multi_host_with_stress.py]: for autotest.stress, when bg.start(), then bg is alive, but stress process is not begin, bg.start() include scp code, make and exec stress, so need wait some seconds for stress process is running.

Comment (xutian) [inline qemu/tests/migration_multi_host_with_stress.py]: so how you ensure sleep time is enough or not enough ? please refer virtio_win related test case to resolve this issue.

thanks,
Xu

**Comment by spcui** [comment]
@xutian updated, could you review again?

File: qemu/tests/cfg/multi_host.cfg [hunk 16]
```diff
@@ -108,6 +108,15 @@
                     #Count of migration during file transfer.
                     migrate_count = 3
                     regain_ip_cmd = ""
+                - with_stress:
+                    only after_login_vm
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: Hi scui,
as we talked, remove before/after_login_vm variants.

File: qemu/tests/cfg/multi_host.cfg [hunk 17]
```diff
@@ -108,6 +108,15 @@
...
+                    only mig_online
```
Comment (suqinhuang) [inline qemu/tests/cfg/multi_host.cfg]: suggest to limit case in mig_offline, then no need to set it for other test any more.

e.g
- mig_offline:
   only with_boot

File: qemu/tests/migration_multi_host_with_stress.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+                    clean_up(vm)
+
+            if params.get("check_vm_before_migration", "yes") == "no":
+                params["check_vm_needs_restart"] = "no"
+
+            self.migrate_wait(self.vms, self.srchost, self.dsthost,
+                              start_worker, check_worker)
+
+    mig = TestMultihostMigration(test, params, env)
+    mig.run()
```
Comment (suqinhuang) [inline qemu/tests/migration_multi_host_with_stress.py]: others looks fine

**Comment by suqinhuang** [comment]
Acked-by: Suqin Huang

**Comment by suqinhuang** [comment]
Acked-by: Suqin Huang

**Pull request merged**